src/erasure-code/isa/ErasureCodeIsa.cc

```diff
@@ -158,7 +190,7 @@ ErasureCodeIsaDefault::isa_decode(int *erasures,
 
   unsigned char *recover_source[k];
   unsigned char *recover_target[m];
-  unsigned char *recover_buf[k+1];
+  char *recover_buf[k+1];
 
```

Comment:
Forgive me if wrong, but if we use signed chars (to represent bytes), when performing arithmetic on a byte with the high bit set then will result in sign extension, right? Pretty sure that in my previous life in FS we used unsigned chars (via macro) to stand for the byte type, which I think is the intention here?

Reply:
The params on isa_encode(char **data, char **coding, int blocksize) are signed, and isa_encode() calls this new isa_xor() function. I don't believe it matters whether the types are signed or unsigned in this case because the xor is performed on the underlying bits of data in the buffers - so it doesn't matter if a bit would represent a sign or not, it will still just be a 1 or a 0 that gets flipped by the xor.

Reply:
Here is a snippet to illustrate the point,  including the output I got (noticed the difference wrt sign)
```c++
#include <cstdio>
#include <cstdint>
#include <iostream>
int main() {
  char a = 0x0a;
  char b = 0x80;
  char c = a ^ b;

  uint8_t x = 0x0a;
  uint8_t y = 0x80;
  uint8_t z = x ^ y;
    std::printf("c: 0x%x\n", c);
    std::printf("z: 0x%x\n", z);
  a = 0xff;
  b = 0x80;
  c = a ^ b;
  x = 0xff;
  y = 0x80;
  z = x ^ y;
    std::printf("c: 0x%x\n", c);
    std::printf("z: 0x%x\n", z);
}// clang++ -Wall -g -std=c++20 -stdlib=libc++ -o xortest xortest.c
// # ./xortest
// c: 0xffffff8a
// z: 0x8a
// c: 0x7f
// z: 0x7f
```

Reply:
The %x format specifier on printf is promoting the chars to ints, hence why you see that output.
Check the values of the variables if I run your program through lldb:
```
Jamies-MacBook-Pro:~ jamie$ lldb ./jose
(lldb) target create "./jose"
Current executable set to '/Users/jamie/jose' (arm64).
(lldb) b jose.cc:12
Breakpoint 1: where = jose`main + 72 at jose.cc:12:30, address = 0x0000000100003ed0
(lldb) run
Process 78657 launched: '/Users/jamie/jose' (arm64)
Process 78657 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x0000000100003ed0 jose`main at jose.cc:12:30
   9   	  uint8_t x = 0x0a;
   10  	  uint8_t y = 0x80;
   11  	  uint8_t z = x ^ y;
-> 12  	    std::printf("c: 0x%x\n", c);
   13  	    std::printf("z: 0x%x\n", z);
   14  	  a = 0xff;
   15  	  b = 0x80;
(lldb) fr v
(char) a = '\n'
(char) b = '\x80'
(char) c = '\x8a'
(uint8_t) x = '\n'
(uint8_t) y = '\x80'
(uint8_t) z = '\x8a'
```

Turns out printf has an %hhx format as well, which is for an unsigned char, so no integer promotion. If I print the values of c using %hhx:
```
Jamies-MacBook-Pro:~ jamie$ ./jose
c: 0x8a
z: 0x8a
c: 0x7f
z: 0x7f
```

Reply:
Yep agree, sorry for the confusion Jamie 👍

Reply:
Approving